File: slp/backend-dotnet/Middlewares/RateLimitingMiddleware.cs
using Microsoft.Extensions.Caching.Distributed;

namespace backend_dotnet.Middlewares;

public class RateLimitingMiddleware
{
    private readonly RequestDelegate _next;
    private readonly ILogger<RateLimitingMiddleware> _logger;
    private const int MaxAttempts = 10;
    private static readonly TimeSpan Window = TimeSpan.FromMinutes(1);

    public RateLimitingMiddleware(RequestDelegate next, ILogger<RateLimitingMiddleware> logger)
    {
        _next = next;
        _logger = logger;
    }

    public async Task Invoke(HttpContext context, IDistributedCache cache)
    {
        // Apply only to login endpoint
        if (context.Request.Path.StartsWithSegments("/api/auth/login") &&
            context.Request.Method == HttpMethods.Post)
        {
            var clientIp = context.Connection.RemoteIpAddress?.ToString() ?? "unknown";
            var key = $"rate:login:{clientIp}";

            var currentCount = await cache.GetStringAsync(key);
            int attemptCount = currentCount == null ? 0 : int.Parse(currentCount);

            if (attemptCount >= MaxAttempts)
            {
                _logger.LogWarning("Rate limit exceeded for IP {Ip}", clientIp);
                context.Response.StatusCode = StatusCodes.Status429TooManyRequests;
                await context.Response.WriteAsync("Too many login attempts. Please try again later.");
                return;
            }

            // Increment counter
            attemptCount++;
            var options = new DistributedCacheEntryOptions
            {
                AbsoluteExpirationRelativeToNow = Window
            };
            await cache.SetStringAsync(key, attemptCount.ToString(), options);

            await _next(context);
        }
        else
        {
            await _next(context);
        }
    }
}

File: slp/backend-nestjs/src/common/middleware/rate-limiting.middleware.ts
import { Injectable, NestMiddleware, HttpException, HttpStatus } from '@nestjs/common';
import { Request, Response, NextFunction } from 'express';
import { Redis } from 'ioredis';
import { ConfigService } from '@nestjs/config';

@Injectable()
export class RateLimitingMiddleware implements NestMiddleware {
  private redis: Redis;
  private readonly maxAttempts = 10;
  private readonly windowSeconds = 60; // 1 minute

  constructor(private configService: ConfigService) {
    const redisUrl = this.configService.get<string>('redis.url');
    if (!redisUrl) throw new Error("No redisUrl!");
    this.redis = new Redis(redisUrl);
  }

  async use(req: Request, res: Response, next: NextFunction) {
    // Apply only to login endpoint
    if (req.path === '/api/auth/login' && req.method === 'POST') {
      const ip = req.ip || req.connection.remoteAddress || 'unknown';

      // Skip rate limit for localhost
      if (ip === '127.0.0.1' || ip === '::1') {
        return next();
      }

      const key = `rate:login:${ip}`;
      const currentCount = await this.redis.get(key);
      const count = currentCount ? parseInt(currentCount, 10) : 0;

      if (count >= this.maxAttempts) {
        throw new HttpException(
          'Too many login attempts. Please try again later.',
          HttpStatus.TOO_MANY_REQUESTS,
        );
      }

      await this.redis.incr(key);
      await this.redis.expire(key, this.windowSeconds);
    }
    next();
  }
}